### DISCUSSION TITLE
Bug in UaExpert dealing with nested struct types

### DISCUSSION THREAD

**Discussion by danmoser** [FreeOpcUa/opcua-asyncio#944] [Q&A] [answered]
When using the UaExpert 1.6.2 (released in March 2022), I started getting the following error:
``Create UaStructureDefinition: Error - field with ValueRank > 1 not supported``

 This can be seen running the [example from asyncua]( https://github.com/FreeOpcUa/opcua-asyncio/blob/master/examples/server-custom-structures-and-enums.py) .

With UaExpert version 1.6.1 (Dez 2021), everything works fine. Screenshots are attached for comparison.

UaExpert 1.6.1:
![nested161](https://user-images.githubusercontent.com/2973004/174913367-f1906906-0d97-40be-9983-99139005be66.png)

UaExpert 1.6.2:
![nested162](https://user-images.githubusercontent.com/2973004/174913376-cb5e4f91-d198-470f-b151-55f563a68a83.png)

**Comment by schroeder-** [comment]
Realy strange in https://reference.opcfoundation.org/v104/Core/docs/Part3/8.51/ I found this:
> The value rank for the field.
> It shall be Scalar (-1) or a fixed rank Array (>=1).

So a ValueRank is legal... Also all our struct only contain ValueRank of 0 or -1 so no clue what ua expert is reading...

Also i tried some things i noticed the following:
- If a struct contains a optional field (StructWithOptionalField), it can't be decoded by uaexpert (older version worked).
- If a struct contains a array with a struct it can not be decoded by uaexpert (older version worked).

**Reply by danmoser** [reply]
Thanks @schroeder- . I opened a support ticket at unified-automation [unified #22793]. I updated this discussion if a have a reply

**Comment by danmoser** [comment, marked as answer]
@schroeder- , I finally got a reply for Unified-Automation:

_A ValueRank of 0 is invalid for structure fields, this is why the log is an error.
General definition of ValueRank values is:_

    Scalar (−1): The value is not an array
    OneDimension (1): The value is an array with one dimension.
    n > 1: the Value is an array with the specified number of dimensions
    OneOrMoreDimensions (0): The value is an array with one or more dimensions
    Any (−2): The value can be a scalar or an array with any number of dimensions
    ScalarOrOneDimension (−3): The value can be a scalar or a one dimensional array.

_These are the possible values for ValueRank on Variables
For structure fields only the following values are valid:_

    Scalar (−1): The value is not an array
    OneDimension (1): The value is an array with one dimension.
    n > 1: the Value is an array with the specified number of dimensions

_But n>1 is new and only supported if the DataType is described through the DataTypeDefinition attribute._

_If a DataTypeDictionary is used to describe a DataType, only Scalar or Array is available._

_The difference between older versions of UaExpert and newer Versions is that we switched from DataTypeDictionary to DataTypeDefinition attribute to be able to display structures._

_If the DataTypeDefinition Attribute has an invalid DataType description (as indicated by the log), it may explain the difference in the display._

Any thoughts?

**Reply by schroeder-** [reply]
Thanks for get in touch with unified-automation.
Easy fix, because it should be OneDimension (1) because we only support lists with 1 dimension, in structs. Fix #1251

**Reply by danmoser** [reply]
The fix #1251 worked here. Thanks, @schroeder- !
